Navier-stokes finite element pieces for AMMT

lindsayad opened the pull request:
Closes #24462

lindsayad committed:
Create INSADConvectedMesh

lindsayad committed:
Update computation of h-max in INSAD for modern AD system

lindsayad committed:
Added strong viscous term support to INSAD

Now comparing equal with INS with all stabilization terms for
PSPG

lindsayad committed:
Regold failing tests

Now that we have the viscous term strong residual, for the RZ cone
tests, I was able to simply set the AD gold file equal to the non-AD
gold file, which is awesome. As I do not have a SUPG energy equation
formulation for INS (non-AD) I did not attempt to do any equivalency
checking for the energy conservation test, and just re-golded the
CSV file

lindsayad committed:
Remove 1e-15 ICs as INSAD should be robust with 0 ICs now

lindsayad committed:
Modernize laser welding 3d input

Closes #24462

lindsayad committed:
Move welding material models from incorrect directory into materials

lindsayad committed:
Create laser welding test

lindsayad committed:
Increase default AD container size

lindsayad committed:
Don't do AD derivative calculations while computing indicators

lindsayad committed:
Remove new nonzero allocations

lindsayad committed:
Remove unused indicators, markers, and postprocessors from input

lindsayad committed:
Give INSAD prefix to new boundary conditions

lindsayad committed:
Rename steel materials for laser welding

lindsayad committed:
Move pieces relevant to laser welding to MOOSE

lindsayad committed:
Documentation

lindsayad committed:
Update for heat conduction bc changes

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1559365/) on 12b9007 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24540/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1559406/) on 12b9007 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2ca47c</th>
      <th colspan=4>#24540 12b900</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.36%</td>
      <td>85.36%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88369</td>
      <td>88372</td>
      <td>+3</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15154</td>
      <td>15154</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24540/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24540/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>2ca47c</th>
      <th colspan=4>#24540 12b900</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.85%</td>
      <td>84.07%</td>
      <td>+0.22%</td>
      <td>95.09%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10785</td>
      <td>10956</td>
      <td>+171</td>
      <td>213</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2078</td>
      <td>2076</td>
      <td>-2</td>
      <td>11</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24540/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24540/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24540/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1560099/) on 79e5e95 : invalidated by @lindsayad

lindsayad commented:
this is ready for review

GiudGiud commented:
on it

GiudGiud reviewed (changes requested):
prefer test or examples? examples will let you locate your code there. 
and it will make it more clear it s not runnable without that code. The whole test object thing only works with the NS module executable

Since the whole domain is fluid I think this can stay in NS. It would be awesome to have the fluid be restricted to the melted region (with an ElementSubdomainModifier maybe?) and have it under FSI

lindsayad commented:
> prefer test or examples? examples will let you locate your code there.
and it will make it more clear it s not runnable without that code. The whole test object thing only works with the NS module executable

Good question. Maybe examples

> Since the whole domain is fluid I think this can stay in NS. It would be awesome to have the fluid be restricted to the melted region (with an ElementSubdomainModifier maybe?) and have it under FSI

Yes I agree that is logical follow-on work. This particular input/model is almost an exact reproduction of the `noble` reference I added where they use the `beta` parameter to do this fuzzy liquid/solid treatment. FSI would be very neat

lindsayad commented on modules/navier_stokes/doc/content/source/kernels/INSADConvectedMesh.md:
```diff
@@ -0,0 +1,21 @@
+where $\rho$ is the density, $\vec{d_m}$ is the fluid mesh displacements, and
+$\vec{u}$ is the fluid velocity. This term is essential for obtaining the
+correct convective derivative of the fluid in cases where the fluid mesh is
+dynamic, e.g. in simulations of fluid-structure interaction.
```
I don't think so because you can do these fuzzy fluid/solid domain approximations. And even if you take the stance that we shouldn't do these fuzzy treatments, I think there's a valid argument for having the object in this module since it adds to the NS momentum equation

GiudGiud replied:
ah I meant to delete that comment but I did not catch it.
This is great in NS

lindsayad commented on modules/navier_stokes/src/materials/INSADMaterial.C:
```diff
@@ -107,6 +108,61 @@ INSADMaterial::subdomainSetup()
+    _disp_x_dot = &disp_x.adUDot();
+    _disp_x_sys_num = disp_x.sys().number();
+    _disp_x_num =
+        disp_x.kind() == (Moose::VarKindType::VAR_NONLINEAR &&
```
wow I feel like the compiler should have errored on this. I guess the boolean is being implicitly converted to an int 😦

lindsayad commented on modules/navier_stokes/test/include/materials/AriaLaserWeld304LStainlessSteelBoundary.h:
```diff
@@ -0,0 +1,64 @@
+ *   year={2007},
+ *   institution={Sandia National Laboratories (SNL), Albuquerque, NM, and Livermore, CA}}
+ */
+class AriaLaserWeld304LStainlessSteelBoundary : public Material
```
No it is just a material that should only be applied in a boundary restricted context because it computes surface quantities like surface tension

lindsayad commented on modules/navier_stokes/test/src/materials/AriaLaserWeld304LStainlessSteel.C:
```diff
@@ -0,0 +1,118 @@
+    _length_units_per_meter(1. / std::pow(10, getParam<int>("length_unit_exponent"))),
+    _temperature_units_per_kelvin(1. / std::pow(10, getParam<int>("temperature_unit_exponent"))),
+    _mass_units_per_kilogram(1. / std::pow(10, getParam<int>("mass_unit_exponent"))),
+    _time_units_per_second(1. / std::pow(10, getParam<int>("time_unit_exponent")))
```
these objects are so old. Definitely opportunities to use some of our neat newer capabilities

lindsayad committed:
Apply suggestions from code review

- Typo fixes

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

AlbertJP7 committed:
Some changes made to some comments.

lindsayad committed:
INSADConvectedMesh -> INSADMeshConvection

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1713365/) on 8aae149 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24540/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24540/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bb8636f71afdaa58e65ae9c616b24dd6e698e733`

lindsayad commented on modules/navier_stokes/doc/content/source/bcs/INSADDisplaceBoundaryBC.md:
```diff
@@ -0,0 +1,22 @@
+previous timestep's value of the displacement, $\Delta t$ is the timestep, and
+$\vec{v}_i$ is a component of the coupled velocity vector, where $i$ denotes the
+component.
+
```
I've implemented an error in the object constructor (will see once I push)

lindsayad commented on modules/navier_stokes/doc/content/source/bcs/INSADDisplaceBoundaryBC.md:
```diff
@@ -0,0 +1,22 @@
+component.
+
+!syntax parameters /BCs/INSADDisplaceBoundaryBC
+
```
I've added some more notes on the possibility of bad elements

lindsayad committed:
Address Guillaume review

- More documentation
- More doxygen
- Clarifying code comments
- Remove unused unit scaling in test materials
- Use `coupledName`
- No need for `const_cast`

lindsayad commented:
> examples will let you locate your code there.

I don't understand. Is there an example of this?

GiudGiud commented:
Tutorials do that (put the code near the inputs)
In terms of examples
tensor_mechanics/examples/uexternaldb_coupling / tensor_mechanics/examples/umat_build do that but it s not common either. 

I think conceptually this is more of an example than a test.

GiudGiud commented:
but it s not shocking for me to have it in tests either. Depends on the visibility you want to have on this.

lindsayad committed:
Move laser welding input from tests to examples

lindsayad commented:
How about something in between? 😄

lindsayad committed:
Remove quotes from block and remove params in markdown

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1717539/) on b5c648b : invalidated by @lindsayad

lindsayad commented:
Ok @GiudGiud I think this should be ready for more review

GiudGiud commented on modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:
```diff
@@ -0,0 +1,43 @@
+    _u_old(_var.nodalValueOld()),
+    _component(getParam<unsigned short>("component"))
+{
+  if (!dynamic_cast<ImplicitEuler *>(_sys.getTimeIntegrator()))
```
btw one can write RK schemes as combinations of Euler schemes so technically we would allow those as well if we are careful.
I think this is fine for now

GiudGiud reviewed (commented):
looks good to me

did you do MMS or other verification of this? It d be worth documenting it

lindsayad commented on modules/navier_stokes/include/bcs/INSADDummyDisplaceBoundaryIntegratedBC.h:
```diff
@@ -0,0 +1,38 @@
+/**
+ * This object adds the sparsity dependence of the surface displacement degrees of
+ * freedom on surface velocity degrees of freedom introduced by the nodal boundary
+ * condition INSADDisplaceBoundaryBC. This sparsity must be added before
```
You want this to be renamed?

GiudGiud replied:
Thinking about it.
Is this displacing the boundary or adapting to a displaced boundary

lindsayad replied:
It is displacing it

lindsayad committed:
Address Guillaume review

- viscuous -> viscous
- Use enum instead of string comparison

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
> did you do MMS or other verification of this? It d be worth documenting it

No, it's just the comparison with the Nobel paper model, which I reference in `laser_welding.md`

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1727901/) on cea60f2 : invalidated by @lindsayad

GiudGiud commented on modules/navier_stokes/include/materials/INSADTauMaterial.h:
```diff
@@ -128,10 +214,112 @@ void
+        _mesh.dimension() >= 2 ? -_mu[_qp] * _d2v[_qp].tr() : ADReal(0);
+    _viscous_strong_residual[_qp](2) =
+        _mesh.dimension() == 3 ? -_mu[_qp] * _d2w[_qp].tr() : ADReal(0);
+    if (_viscous_form == "traction")
```
above & beyond

lindsayad commented:
@GiudGiud anything remaining? No rush

lindsayad commented:
Maybe a low simmer of a rush now. The AMMT report is due to PRS in a couple weeks and it would be nice to point to a merged pull request if possible

GiudGiud commented:
Writing my part of TA report tomorrow then I'll get on this

GiudGiud reviewed (commented):
did a full sweep, done after that

lindsayad committed:
Address Guillaume review

- Add note about test objects in laser welding example
- Clarifying comments for
    - Viscous strong residual calculation
    - Mesh convection
- Add another reference regarding the mesh convection term
- Add `Moving mesh` parameter group

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on modules/navier_stokes/examples/laser-welding/gold/3d_exodus.e-s004:
these gold files have 0 velocity, 0 pressure, and exceedingly small displacement.
only T is moving. can we explain it?

I havent run at 2x refinement, just the test conditions (and played with parameters there)

lindsayad replied:
The tests spec only runs a smidgen of the simulation in order to verify syntax and ability to solve. If you run without the tests spec command line arguments then you will get a complete simulation with actual stuff happening.

For a while all that is significantly changing is the condensed phase temperature. Then after a while the vapor pressure gets high enough and you have significant evaporation which imposes a momentum recoil force on the surface. Then things start to move and displace. Surface tension forces would also add an effect

lindsayad commented:
Thanks for the reviews @GiudGiud
